### ISSUE TITLE
Property `app_name` should be removed from `Device` type definition.

### ISSUE THREAD

**Issue by srlowe** [balena-io/balena-sdk#884]
See: https://www.flowdock.com/app/rulemotion/resin-frontend/threads/DXXSZ4-_EXcXp6b1-WxZ1G4cyXO

**Comment by tomgs** [comment]
I was literally about to open a PR mentioning that I'd like to have that as a property inside a device. I now manually map between the ID and the name using an additional API call. What is the reasoning behind getting rid of it?

**Comment by thgreasi** [comment]
I actually had forgotten about this :sweat_smile: 
It should be part of the next major release.

@tomgs the reasoning is that the `app_name` field isn't part of the actual Device model.
This field was mistakenly added in the Device typings and should be removed.

You can see the device model in the following pieces of code of the open-balena-api:
See: https://github.com/balena-io/open-balena-api/blob/72516a8aa58da5e807e6c164b1c1892f4b4a642d/src/balena.sbvr#L285-L314
See: https://github.com/balena-io/open-balena-api/blob/72516a8aa58da5e807e6c164b1c1892f4b4a642d/src/balena.sbvr#L420-L487

> I'd like to have that as a property inside a device.

Can you clarify whether you just want the typings or the field being actually populated?

**Comment by tomgs** [comment]
@thgreasi that makes sense, but still - costs me 2 API calls instead of one:)
And specifically, the fact that the App **ID** is not part of the device model is absurd to me.
Given a device object pulled from the API (I'm using the `nodejs` SDK now, actually, but I assume feature parity) the device's app ID is nested, for some reason, in here:

```json
"device (belongs_to__application" : { 
    __deferred: 
        { uri: '/resin/application(1040245)' }, 
__id: 1040245 }
```
Looking at the typings:

https://github.com/balena-io/balena-sdk/blob/master/typings/balena-sdk.d.ts#L510

It seems like the `belongs_to__application` property is generic, which kinda explains this weird nesting bit (I think), but it still feels a bit... awkward to me.

Anywho, I might be just ranting for sport by now. This is an architectural decision that I was not a part of, and I'm missing some context to be of service. Will go back to my 2 API calls... :)

Thanks @thgreasi for the quick response!

**Comment by thgreasi** [comment]
@tomgs our backend actually does support expanding towards related resources with a graphql like approach. For example you can see in the implementation of `device.getApplicationName` that we only need to do a single request:
See: https://github.com/balena-io/balena-sdk/blob/b417df50c20de8629208aff5ef56f5be3b2d1082/lib/models/device.js#L608-L614

You can use `$expand` on all properties that are typed as NavigationResources & ReverseNavigationResources.
For example you can do something like:
```js
const onlineDevices = await sdk.models.device.getAll({
	$select: [
		'id',
		'is_online',
		'device_name',
		'os_variant',
		'uuid',
	],
	$expand: {
		belongs_to__application: {
			$select: [
				'id',
				'app_name',
				'application_type',
			],
		},
	},
	$filter: {
		is_online: true,
	},
});

onlineDevices.forEach((d) => {
	console.log(
		d.uuid,
		d.device_name,
		d.belongs_to__application[0].app_name,
	);
});
```

For properties that are Foreign Keys and are not part of a `$expand` we indeed by default populate them with an object like the one you mentioned above, so that consumer code can easily reason that this property is pointing to a different resource and is not just a plain integer.

**Comment by tomgs** [comment]
You don't say. Away from the machine now, back in a bit and will test it out. Thanks for pointing this out!

**Comment by tomgs** [comment]
Oh, this is a beautiful approach. Well done!

I have this neat little function now:

```js

import * as initBalena from "balena-sdk";

const dataDirectory = process.env.DATA_DIRECTORY;
const balena = initBalena({ dataDirectory });

const getAllBalenaDevices = async () : Promise<initBalena.Device[]> => {
  return await balena.models.device.getAll({
    $select: [
      'uuid',
      'last_connectivity_event',
      'device_name'
    ],
    $expand: {
      belongs_to__application: {
        $select: [
          'app_name',
        ],
      },
    }
  })
}
```

That gets me all the relevant information.

**Comment by thgreasi** [comment]
Resolved by #621

**Comment by tomgs** [comment]
@thgreasi not sure I follow. But thanks!

**Comment by thgreasi** [comment]
@tomgs I mean that we indeed completed what this issue was describing and remove the app_name field from the Device typings in the just released balena-sdk v14.0.0.

Now about the returned type from `application.getAll()`, indeed we currently do not return exact types, so the returned object is `PineDeferred | Application[]`.
We do intend to properly support returning exact types from all calls and have done some work on this already.
See: https://github.com/balena-io/balena-sdk/issues/887
See: https://github.com/balena-io/balena-sdk/pull/889

Once that's in place, TypeScript will be seeing whether you have passed in a `belongs_to__application` expand and type the respective property of the result as `Application[]`.
For now, if you know that you have expanded a specific property you can use a helper noop function to fix the typings. We are internally using something like:
```ts
export const getExpanded = <T>(obj: BalenaSdk.NavigationResource<T>) =>
	Array.isArray(obj) ? obj[0] : undefined;

const appName = getExpanded(device.belongs_to__application)!.app_name;
```
